Add 'spo group user add' command. Closes #2483

arjunumenon opened the pull request:
> ### Create a new Command 'spo group user add' for adding the user to SharePoint Group
>
>Adds 'spo group user add' command for adding the user to SharePoint Group. Closes #2483

> ### Linked Issue

Closes #2483

arjunumenon commented:
@pnp/cli-for-microsoft-365-maintainers  - Just an FYI about the implementation. For adding the users to SharePoint group, we have used the endpoint, `/_api/SP.Web.ShareObject` rather than the conventional endpoint, `_api/web/SiteGroups/GetById(...)/Users/AddUserById(...)`. 

The reason is because, our command gives the feature to add multiple users to the group. If we go with the latter endpoint, it does not support the addition of multiple users in the single call, while with former, you can add multiple users in the single command and our code does not need to handle looping mechanism.

Let me know if you see any challenges in the approach.

waldekmastykarz commented:
If it works, it works 👏

waldekmastykarz commented on docs/docs/cmd/spo/group/group-user-add.md:
```diff
@@ -0,0 +1,36 @@
+`-u, --webUrl <webUrl>`
+: URL of the site where the SharePoint group is available
+
+`-i, --groupId <groupId>`
```
I didn't catch it in the spec, but let's remove the short `-i` here. It would be ambiguous if we ever extended the command with support for specifying `userId`.

arjunumenon replied:
Sure. That make sense considering the extension of the command.
Have updated the specs as well so that [specs ](https://github.com/pnp/cli-microsoft365/issues/2483#issue-916289528)are consistent.

waldekmastykarz commented on docs/docs/cmd/spo/group/group-user-add.md:
```diff
@@ -0,0 +1,36 @@
+`-i, --groupId <groupId>`
+: Id of the SharePoint Group to which user needs to be added
+
+`-n, --userName <userName>`
```
I didn't catch it in the spec, but let's remove the short `-n` here. It would be ambiguous if we ever extended the command with support for specifying `groupName`.

arjunumenon replied:
Sure. Have updated the [specs](https://github.com/pnp/cli-microsoft365/issues/2483#issue-916289528)

waldekmastykarz commented on src/m365/spo/commands/group/group-user-add.ts:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    const data: any = {
+      url: `${args.options.webUrl}`,
```
```suggestion
      url: args.options.webUrl,
```

No need to interpolate the string as we pass it as-is

waldekmastykarz commented on src/m365/spo/commands/group/group-user-add.ts:
```diff
@@ -0,0 +1,111 @@
+    };
+
+    request.post<string>(requestOptions)
+      .then((rawRes: any): void => {
```
As we expect the returned object to have properties, let's define them on the returned type rather than using `any`, so that we can avoid errors in refactoring in the future.

waldekmastykarz commented on src/m365/spo/commands/group/group-user-add.ts:
```diff
@@ -0,0 +1,111 @@
+      responseType: 'json'
+    };
+
+    request.post<string>(requestOptions)
```
The returned object isn't a string but a JSON object with properties, so let's properly define it on the `post()` method

waldekmastykarz commented on src/m365/spo/commands/group/group-user-add.ts:
```diff
@@ -0,0 +1,111 @@
+        if (rawRes.StatusCode !== 0) {
+          return cb(new CommandError(rawRes.ErrorMessage));
+        }
+        logger.log(rawRes.UsersAddedToGroup);
```
If you specify a non-existent user name, the command fails with an unhandled exception because the `rawRes.UsersAddedToGroup` property is null. We should check for this in our code.

arjunumenon replied:
That is weird. When I am executing the username which does not exist, I am getting "Couldn't resolve user" in the errormessage property. 🤔.
![image](https://user-images.githubusercontent.com/10369964/123596378-c10e9700-d80f-11eb-8921-b7723e9db327.png)

Not sure whether I am overlooking anything else in the code.

waldekmastykarz replied:
Thinking of it, I wonder if it wouldn't be better to check first if the specified user exists, to avoid creating unnecessary invites that will clutter the UI. What do you think @arjunumenon?

arjunumenon replied:
Good point when you consider the links getting cluttered. Validation check make sense to avoid that situation.

Just to understand, What is in your mind on how to handle that. Do you prefer to,

1. Have a validation check before the command gets executed. So if there are any users who do not exist, command will not be executed and validation fails
2. Or when you execute the command, we would only consider the users which exists currently.

For usability point of view, **No 2** would be preferable. But do you think it would burden the command execution? When we have multiple users in the list of users, we may have to check each users during the execution phase and I was wondering it may impact the execution time. Pardon me, just playing Devil's Advocate here. 😈. 

Your thoughts please @waldekmastykarz ?

waldekmastykarz replied:
Unfortunately, right now we can't execute a validation request in the validation stage because then the auth hasn't been restored yet. So the only option we have available is to perform the necessary checks while executing the command.

arjunumenon replied:
Thanks Waldek. In that case, what I was thinking is to have the check before executing the command and pass only the active / resolved users to the `Sharing` Method. Hope that approach works.

Another question on the same line. Since we already have a CLI command which gets the username, 

- Can we call another command from one command. Is that a recommended approach?
- Or do you prefer to make another REST call which checks the user?

waldekmastykarz replied:
We can totally call another command. If it does exactly what you need, it would help us centralize the code. To do that, you'd use the `Cli.executeCommandWithOutput` method: https://github.com/pnp/cli-microsoft365/blob/04b66058b775750239a2b18dbc46634eba86257d/src/cli/Cli.ts#L233

arjunumenon replied:
That is awesome. Had not tried that before. Let me try that and will see if it works as expected. Should work. Will update the code accordingly.

waldekmastykarz commented on src/m365/spo/commands/group/group-user-add.ts:
```diff
@@ -0,0 +1,111 @@
+      }, (rawRes: any): void => this.handleRejectedODataJsonPromise(rawRes, logger, cb));
+  }
+
+  private getFormatteduserList(userList: string): string {
```
```suggestion
  private getFormattedUserList(userList: string): string {
```

waldekmastykarz commented on src/m365/spo/commands/group/group-user-add.ts:
```diff
@@ -0,0 +1,111 @@
+  }
+
+  private getFormatteduserList(userList: string): string {
+    const splittedUserName: string[] = userList.split(",");
```
`splitted` is not a word. The proper word is `split`. Let's adjust it throughout 😊

arjunumenon replied:
Aah. nice catch. English is hard at times.. 🙄

waldekmastykarz commented on src/m365/spo/commands/group/group-user-add.ts:
```diff
@@ -0,0 +1,111 @@
+    const splittedUserName: string[] = userList.split(",");
+    let generatedPeoplePicker: string = "[";
+
+    splittedUserName.forEach((userName) => {
```
Rather than manually concatenating the string and escaping JSON strings, let's refactor this using `.map` and JSON.stringify. Using `.map` we can turn each `userName` into an object, then serialize it using `JSON.stringify` and finally join the array of strings into a ,-separated string and wrap in `[...]`.

waldekmastykarz reviewed (changes requested):
We're almost there. Let's do a few adjustments before we merge it.

arjunumenon commented:
> We're almost there. Let's do a few adjustments before we merge it.

Thanks Waldek for the detailed review. It really shows that I have become rusty in CLI 🙄. Post the code review  is the best part of me with your valuable comments. Thanks for being patient in reviewing.

Will look into the comments and update you.

waldekmastykarz commented:
No worries, we're in this together 👍

waldekmastykarz commented:
Thanks! Let's do another pass 👍

arjunumenon commented:
Hey Waldek - I have made couple of changes in the following PR.

- Validation of active users before initiating adding to the groups
For this. we have used the consumption of existing command as per [this conversation](https://github.com/pnp/cli-microsoft365/pull/2530#discussion_r664432173). 
- Removed interfaces which are not used for the functionality.

Like always looking forward your valuable feedback.

arjunumenon commented:
Ready to initiate PR. Since there is lack of 100% coverage for the main branch as per #2586 , keeping this on hold. Once that is complete, will initiate the PR after rebasing the fix.

arjunumenon commented:
Marking the PR **Ready for Review** since #2586 is already closed.

waldekmastykarz reviewed (changes requested):
When trying to add a valid user, I'm getting an error: `Error: Couldn't resolve the users.`. The API call that's failing is:

```text
GET https://m365x954810.sharepoint.com/sites/contoso/_api/web/siteusers/GetByEmail('AdeleV%40M365x954810.OnMicrosoft.com')
```

```text
{
  "status": 500,
  "statusText": "Internal Server Error",
  "headers": {
    ...
  },
  "error": {
    "odata.error": {
      "code": "-2146232832, Microsoft.SharePoint.SPException",
      "message": {
        "lang": "en-US",
        "value": "User cannot be found."
      }
    }
  }
}
```

I suspect that the issue is caused by the user existing AAD but not being added to the site yet. We should adjust our validation logic to take this scenario into account.

arjunumenon commented:
> When trying to add a valid user, I'm getting an error: Error: Couldn't resolve the users.. The API call that's failing is:

Ahh that is a scenario which I had not foreseen. Sorry that you had review again without success. Silly, but huge mistake. 🤦‍♂️
Thanks for finding the root cause too. 👍

I was using the command `spo user get` for validating the user and for that user to be part of the User Information List. But in this case, user just need to be part of AAD. Hence, I have changed the command used to be `aad user get` which will ensure that user do exists.

I had implemented the changes and had tried various test cases to ensure that we are not missing anything. I will be initiating PR shortly with the changes. 
Can you let me know if I am overlooking anything else.

waldekmastykarz commented:
No worries, that's why we check each other's work to ensure that we haven't missed anything. Thank you for a quick turnaround 👏

garrytrinder reviewed (changes requested):
Good work @arjunumenon however I'm wondering if we need to make a change to how the multiple user error handling works.

See below example, I am passing valid and non valid users however no error is presented for the user that does not exist, which leads to a false positive.

<img width="613" alt="image" src="https://user-images.githubusercontent.com/11563347/128665935-4963dc17-1b47-4c36-ac2e-fe4409bbf87d.png">

If I was to use this command to bulk add users to a group, then I would most likely want to know which users did not get successfully added.

arjunumenon commented:
Thanks @garrytrinder for the review. It is definitely a case where you will get a false positive if in actual outcome is partial success..
But I was not sure how did we show it across to the end users. 

Do you envision to  show the failure users as a different message along with the added users? If that is the case, we may have to alter the default output object of the API call.
I cannot think of other commands where we can have partial success in the single command. Any thoughts on that please.?

garrytrinder commented:
> I cannot think of other commands where we can have partial success in the single command. Any thoughts on that please.?

A similar command to this would be `aad o365group set` which accepts multiple owners, looking at its implementation just appear to be validating that each user string is a valid format.

https://github.com/pnp/cli-microsoft365/blob/928c2ed0b4a6c6f7771178de7f52ac511a6bc411/src/m365/aad/commands/o365group/o365group-set.ts#L262-L269

However, the same test passing, i.e. an incorrect user, results in no error message being displayed.

<img width="827" alt="image" src="https://user-images.githubusercontent.com/11563347/128977874-7bb96e2f-c5c4-4860-966c-46bc68b38eb3.png">

As your command exhibits the same behaviour, shall we create a discussion to continue this conversation separate to this PR, we can then get this merged and come back to the error handling as a feature enhancement?

arjunumenon commented:
Thanks Garry for looking into that. I guess handling **partial success** scenarios for multiple values are not quite common across. 
In parallel, I had checked some of the PowerShell commands and somehow I could not find commands which takes care of this kind of scenarios.

Good idea to create a discussion and we can brainstorm and decide and can update the commands accordingly.

garrytrinder commented:
Merged manually 👍
